ProcessLib/SmallDeformationWithLIE/Common/MeshUtils.cpp

```diff
@@ -62,73 +63,106 @@ class IsCrackTip
+        std::vector<std::vector<MeshLib::Element*>>& vec_fracture_matrix_elements,
+        std::vector<std::vector<MeshLib::Node*>>& vec_fracture_nodes
         )
 {
     IsCrackTip isCrackTip(mesh);
 
     // get vectors of matrix elements and fracture elements
     vec_matrix_elements.reserve(mesh.getNumberOfElements());
+    std::vector<MeshLib::Element*> all_fracture_elements;
     for (MeshLib::Element* e : mesh.getElements())
     {
         if (e->getDimension() == mesh.getDimension())
             vec_matrix_elements.push_back(e);
         else
-            vec_fracture_elements.push_back(e);
+            all_fracture_elements.push_back(e);
     }
     DBUG("-> found total %d matrix elements and %d fracture elements",
-         vec_matrix_elements.size(), vec_fracture_elements.size());
+         vec_matrix_elements.size(), all_fracture_elements.size());
+
+    // get fracture material IDs
+    auto opt_material_ids(mesh.getProperties().getPropertyVector<int>("MaterialIDs"));
+    if (!opt_material_ids)
+        OGS_FATAL("MaterialIDs propery vector not found in a mesh");
+    for (MeshLib::Element* e : all_fracture_elements)
+        vec_fracture_mat_IDs.push_back((*opt_material_ids)[e->getID()]);
+    BaseLib::makeVectorUnique(vec_fracture_mat_IDs);
+    DBUG("-> found %d fracture material groups", vec_fracture_mat_IDs.size());
+
+    // create a vector of fracture elements for each group
+    vec_fracture_elements.resize(vec_fracture_mat_IDs.size());
+    for (unsigned frac_id=0; frac_id<vec_fracture_mat_IDs.size(); frac_id++)
+    {
+        const auto frac_mat_id = vec_fracture_mat_IDs[frac_id];
+        std::vector<MeshLib::Element*> &vec_elements = vec_fracture_elements[frac_id];
+        std::copy_if(all_fracture_elements.begin(), all_fracture_elements.end(),
+                     std::back_inserter(vec_elements),
+                     [&](MeshLib::Element* e) {
+                         return (*opt_material_ids)[e->getID()] == frac_mat_id;
+                     });
+        DBUG("-> found %d elements on the fracture %d", vec_elements.size(), frac_id);
+    }
 
     // get a vector of fracture nodes
-    for (MeshLib::Element* e : vec_fracture_elements)
+    vec_fracture_nodes.resize(vec_fracture_mat_IDs.size());
+    for (unsigned frac_id=0; frac_id<vec_fracture_mat_IDs.size(); frac_id++)
     {
-        for (unsigned i=0; i<e->getNumberOfNodes(); i++)
+        std::vector<MeshLib::Node*> &vec_nodes = vec_fracture_nodes[frac_id];
+        for (MeshLib::Element* e : vec_fracture_elements[frac_id])
         {
-            if (isCrackTip(*e->getNode(i)))
-                continue;
-            vec_fracture_nodes.push_back(const_cast<MeshLib::Node*>(e->getNode(i)));
+            for (unsigned i=0; i<e->getNumberOfNodes(); i++)
+            {
+                if (isCrackTip(*e->getNode(i)))
+                    continue;
+                vec_nodes.push_back(const_cast<MeshLib::Node*>(e->getNode(i)));
```

Comment:
Just wondering: Can't the nodes be `const`? (Then we probably need `std::vector<std::vector<MeshLib::Node const*>>& vec_fracture_nodes`, too)